module/zfs/bpobj.c

```diff
@@ -264,97 +306,176 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
+				}
+			}
+			/*
+			 * If there are no entries, there should
+			 * be no bytes.
+			 */
+			if (bpobj_is_empty(bpo)) {
+				ASSERT0(bpo->bpo_phys->bpo_bytes);
+				ASSERT0(bpo->bpo_phys->bpo_comp);
+				ASSERT0(bpo->bpo_phys->bpo_uncomp);
+			}
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			continue;
+		}
+
+		ASSERT(bpo->bpo_havecomp);
+
+		dmu_object_info_t doi;
+		err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs,
+		    &doi);
+		if (err) {
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			mutex_exit(&bpo->bpo_lock);
+			return (err);
+		}
+		ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
+		int epb = doi.doi_data_block_size / sizeof (uint64_t);
+
+		/*
+		 * Can't clean up bpobj while it still has subobjs.  Place it
+		 * back in the list ahead of its children.  It will be cleaned
+		 * up once the children are all freed.
+		 */
+		if (bpo->bpo_phys->bpo_num_subobjs > 0)
+			list_insert_head(&stack, bpi);
 
+		/* Add the last subobj to stack. */
+		int64_t i = bpo->bpo_phys->bpo_num_subobjs - 1;
+		uint64_t offset = i * sizeof (uint64_t);
+		uint64_t blkoff = P2PHASE(i, epb);
 		if (dbuf == NULL || dbuf->db_offset > offset) {
 			if (dbuf)
 				dmu_buf_rele(dbuf, FTAG);
 			err = dmu_buf_hold(bpo->bpo_os,
-			    bpo->bpo_phys->bpo_subobjs, offset, FTAG, &dbuf, 0);
+			    bpo->bpo_phys->bpo_subobjs, offset, FTAG,
+			    &dbuf, 0);
 			if (err)
 				break;
 		}
 
 		ASSERT3U(offset, >=, dbuf->db_offset);
 		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
 
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
-			}
-		}
-		err = bpobj_iterate_impl(&sublist, func, arg, tx, free);
-		if (free) {
-			VERIFY3U(0, ==, bpobj_space(&sublist,
-			    &used_after, &comp_after, &uncomp_after));
-			bpo->bpo_phys->bpo_bytes -= used_before - used_after;
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			bpo->bpo_phys->bpo_comp -= comp_before - comp_after;
-			bpo->bpo_phys->bpo_uncomp -=
-			    uncomp_before - uncomp_after;
-		}
+		uint64_t *objarray = dbuf->db_data;
+		bpobj_t *sublist = kmem_alloc(sizeof (bpobj_t),
+		    KM_SLEEP);
 
-		bpobj_close(&sublist);
+		err = bpobj_open(sublist, bpo->bpo_os, objarray[blkoff]);
 		if (err)
 			break;
-		if (free) {
-			err = dmu_object_free(bpo->bpo_os,
-			    objarray[blkoff], tx);
-			if (err)
-				break;
-			bpo->bpo_phys->bpo_num_subobjs--;
-			ASSERT3S(bpo->bpo_phys->bpo_num_subobjs, >=, 0);
+		if (free)
+			dmu_buf_will_dirty(sublist->bpo_dbuf, tx);
+
+		list_insert_head(&stack, bpi_alloc(sublist, bpi,
+		    objarray[blkoff], i));
+
+		if (dbuf) {
+			dmu_buf_rele(dbuf, FTAG);
+			dbuf = NULL;
```

Comment:
Since you're doing this, dbuf is not actually cached between each iteration of the loop.  In that case you can simplify the code by eliminating `dbuf` (and calls to dmu_buf_hold/rele) and instead use dmu_read() to read the 8 bytes we need out of `bpo_subobjs`

Reply:
Nice.  This eliminates a bunch of code including the call to dmu_object_info since we no longer need to get the epb.